File: blind75/06_maximum_product_subarray/maximum_product_subarray.py
from typing import List


def max_product(nums: List[int]) -> int:
    """
    Finds the contiguous subarray with the largest product.
    Uses modified Kadane's algorithm tracking both max and min products.

    Args:
        nums: List of integers

    Returns:
        The maximum product of a contiguous subarray
    """
    if not nums:
        return 0

    max_so_far = nums[0]
    max_ending_here = nums[0]
    min_ending_here = nums[0]

    for i in range(1, len(nums)):
        num = nums[i]

        # Calculate candidates: current number, max*num, min*num
        # We need temp variable because max_ending_here is used in min_ending_here calculation
        temp_max = max(num, num * max_ending_here, num * min_ending_here)
        min_ending_here = min(num, num * max_ending_here, num * min_ending_here)
        max_ending_here = temp_max

        max_so_far = max(max_so_far, max_ending_here)

    return max_so_far

File: blind75/06_maximum_product_subarray/maximum_product_subarray.go
package maximum_product_subarray

// MaxProduct finds the contiguous subarray with the largest product.
// Uses modified Kadane's algorithm tracking both max and min products.
func MaxProduct(nums []int) int {
	if len(nums) == 0 {
		return 0
	}

	maxSoFar := nums[0]
	maxEndingHere := nums[0]
	minEndingHere := nums[0]

	for i := 1; i < len(nums); i++ {
		num := nums[i]

		// Calculate candidates: current number, max*num, min*num
		// We need temp variable because maxEndingHere is used in minEndingHere calculation
		tempMax := max(num, max(num*maxEndingHere, num*minEndingHere))
		minEndingHere = min(num, min(num*maxEndingHere, num*minEndingHere))
		maxEndingHere = tempMax

		maxSoFar = max(maxSoFar, maxEndingHere)
	}

	return maxSoFar
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func min(a, b int) int {
	if a < b {
		return a
	}
	return b
}
